File: src/stores/journal.ts
import { defineStore } from "pinia";
import { invoke } from "@tauri-apps/api/core";
import { open, save } from "@tauri-apps/plugin-dialog";
import type { ExportPreview, JournalEntry, TemplateRecord } from "../types";
import { useAppStore } from "./app";

export const useJournalStore = defineStore("journal", {
  state: () => ({
    entries: [] as JournalEntry[],
    templates: [] as TemplateRecord[],
    current: null as JournalEntry | null,
    preview: null as ExportPreview | null,
    loading: false,
  }),
  actions: {
    async refresh() {
      const app = useAppStore();
      try {
        const [entries, templates] = await Promise.all([
          invoke<JournalEntry[]>("list_journal_entries"),
          invoke<TemplateRecord[]>("list_templates"),
        ]);
        this.entries = entries;
        this.templates = templates;
      } catch (e) {
        app.setError(String(e));
      }
    },

    async loadEntry(id: string) {
      const app = useAppStore();
      try {
        this.current = await invoke<JournalEntry>("get_journal_entry", { id });
      } catch (e) {
        app.setError(String(e));
      }
    },

    async create(input: {
      title: string;
      template_id?: string | null;
      body_md: string;
      fields?: Record<string, unknown>;
      tags?: string[];
      entry_date?: string;
    }) {
      const app = useAppStore();
      this.loading = true;
      try {
        const entry = await invoke<JournalEntry>("create_journal_entry", {
          input: {
            title: input.title,
            template_id: input.template_id ?? null,
            body_md: input.body_md,
            fields: input.fields ?? {},
            tags: input.tags ?? [],
            entry_date: input.entry_date ?? null,
          },
        });
        this.current = entry;
        await this.refresh();
        return entry;
      } catch (e) {
        app.setError(String(e));
        return null;
      } finally {
        this.loading = false;
      }
    },

    async update(
      id: string,
      input: {
        title: string;
        template_id?: string | null;
        body_md: string;
        fields?: Record<string, unknown>;
        tags?: string[];
        entry_date?: string;
      },
    ) {
      const app = useAppStore();
      this.loading = true;
      try {
        const entry = await invoke<JournalEntry>("update_journal_entry", {
          id,
          input: {
            title: input.title,
            template_id: input.template_id ?? null,
            body_md: input.body_md,
            fields: input.fields ?? {},
            tags: input.tags ?? [],
            entry_date: input.entry_date ?? null,
          },
        });
        this.current = entry;
        await this.refresh();
        return entry;
      } catch (e) {
        app.setError(String(e));
        return null;
      } finally {
        this.loading = false;
      }
    },

    async remove(id: string) {
      const app = useAppStore();
      try {
        await invoke("delete_journal_entry", { id });
        if (this.current?.id === id) this.current = null;
        await this.refresh();
      } catch (e) {
        app.setError(String(e));
      }
    },

    async previewExport(id: string) {
      const app = useAppStore();
      try {
        this.preview = await invoke<ExportPreview>("preview_journal_export", { id });
      } catch (e) {
        app.setError(String(e));
      }
    },

    async exportToFile(id: string) {
      const app = useAppStore();
      try {
        await this.previewExport(id);
        const path = await save({
          filters: [{ name: "Markdown", extensions: ["md"] }],
          defaultPath: `${this.preview?.title || "entry"}.md`,
        });
        if (!path) return;
        await invoke("export_journal_entry", { id, path });
      } catch (e) {
        app.setError(String(e));
      }
    },

    async createTemplate(input: {
      name: string;
      description?: string;
      category?: string;
      fields: unknown[];
      body_md: string;
      default_tags?: string[];
    }) {
      const app = useAppStore();
      try {
        await invoke("create_template", { input });
        await this.refresh();
      } catch (e) {
        app.setError(String(e));
      }
    },

    async saveEntryAsTemplate(entryId: string, name: string) {
      const app = useAppStore();
      try {
        await invoke("save_entry_as_template", { entryId, name });
        await this.refresh();
      } catch (e) {
        app.setError(String(e));
      }
    },

    async openInChroma(spectrumPath: string) {
      const app = useAppStore();
      try {
        await invoke("open_in_chroma", {
          spectrumPath,
          range: null,
        });
      } catch (e) {
        app.setError(String(e));
      }
    },

    async exportTemplateFile(id: string) {
      const app = useAppStore();
      try {
        const path = await save({
          filters: [{ name: "SoheiDesk template", extensions: ["json"] }],
          defaultPath: "template.json",
        });
        if (!path) return;
        await invoke("export_template_file", { id, path });
      } catch (e) {
        app.setError(String(e));
      }
    },

    async importTemplateFile() {
      const app = useAppStore();
      try {
        const path = await open({
          multiple: false,
          filters: [{ name: "JSON", extensions: ["json"] }],
        });
        if (typeof path !== "string") return;
        await invoke("import_template_file", { path });
        await this.refresh();
      } catch (e) {
        app.setError(String(e));
      }
    },

    clearPreview() {
      this.preview = null;
    },
  },
});

File: src-tauri/src/commands/journal.rs
use crate::db::DbState;
use crate::error::AppResult;
use crate::journal::{self, ExportPreview, JournalEntry, JournalEntryInput};
use crate::search::{self, SearchState};
use crate::templates::{self, TemplateInput, TemplateRecord};
use tauri::State;

#[tauri::command]
pub fn list_templates(db: State<'_, DbState>) -> AppResult<Vec<TemplateRecord>> {
    templates::list_templates(&db)
}

#[tauri::command]
pub fn create_template(db: State<'_, DbState>, input: TemplateInput) -> AppResult<TemplateRecord> {
    templates::create_template(&db, input)
}

#[tauri::command]
pub fn update_template(
    db: State<'_, DbState>,
    id: String,
    input: TemplateInput,
) -> AppResult<TemplateRecord> {
    templates::update_template(&db, &id, input)
}

#[tauri::command]
pub fn delete_template(db: State<'_, DbState>, id: String) -> AppResult<()> {
    templates::delete_template(&db, &id)
}

#[tauri::command]
pub fn list_journal_entries(db: State<'_, DbState>) -> AppResult<Vec<JournalEntry>> {
    journal::list_entries(&db)
}

#[tauri::command]
pub fn get_journal_entry(db: State<'_, DbState>, id: String) -> AppResult<JournalEntry> {
    journal::get_entry(&db, &id)
}

#[tauri::command]
pub fn create_journal_entry(
    db: State<'_, DbState>,
    search: State<'_, SearchState>,
    input: JournalEntryInput,
) -> AppResult<JournalEntry> {
    let entry = journal::create_entry(&db, input)?;
    let _ = search::index_journal_entry(&search, &entry.id, &entry.title, &entry.body_md);
    Ok(entry)
}

#[tauri::command]
pub fn update_journal_entry(
    db: State<'_, DbState>,
    search: State<'_, SearchState>,
    id: String,
    input: JournalEntryInput,
) -> AppResult<JournalEntry> {
    let entry = journal::update_entry(&db, &id, input)?;
    let _ = search::index_journal_entry(&search, &entry.id, &entry.title, &entry.body_md);
    Ok(entry)
}

#[tauri::command]
pub fn delete_journal_entry(
    db: State<'_, DbState>,
    search: State<'_, SearchState>,
    id: String,
) -> AppResult<()> {
    journal::delete_entry(&db, &id)?;
    let _ = search.delete(&id);
    Ok(())
}

#[tauri::command]
pub fn preview_journal_export(db: State<'_, DbState>, id: String) -> AppResult<ExportPreview> {
    journal::preview_export(&db, &id)
}

#[tauri::command]
pub fn export_journal_entry(
    db: State<'_, DbState>,
    id: String,
    path: String,
) -> AppResult<()> {
    journal::export_entry_to_path(&db, &id, &path)
}

#[tauri::command]
pub fn save_entry_as_template(
    db: State<'_, DbState>,
    entry_id: String,
    name: String,
) -> AppResult<TemplateRecord> {
    journal::save_entry_as_template(&db, &entry_id, name)
}

#[tauri::command]
pub fn export_template_file(
    db: State<'_, DbState>,
    id: String,
    path: String,
) -> AppResult<()> {
    templates::export_template_to_path(&db, &id, &path)
}

#[tauri::command]
pub fn import_template_file(db: State<'_, DbState>, path: String) -> AppResult<TemplateRecord> {
    templates::import_template_from_path(&db, &path)
}
